### PR TITLE
Support custom types

### LINKED ISSUE #226: `type_<fun()>`

### LINKED ISSUE #227: `"lm"`

### LINKED ISSUE #228: `"glm"`

### LINKED ISSUE #229: `"loess"`

### PR DESCRIPTION
**Pull request by vincentarelbundock**
This is the last preparatory step in my master plan to allow custom user-supplied geoms.

Split `draw_elements()` into a separate drawing function for each geom, such as `draw_polygon()`, `draw_ribbon()`, etc. Then, `draw_elements()` uses a `switch()` call to select the correct one based on the value  of `type`.

I wrote this from a fork of your `flip` PR, so it can supplant it.

Sorry, I should have waited for a merge, but the family went away for a thing and I found myself with a couple spare hours.

_GM edit/update for tracking purposes:_

Closes #226, closes #227, closes #228, closes #229.

### CODE REVIEW

**Comment by grantmcdermott** [comment]
Excellent, thanks! I'll take a proper look at this once AZ has had a chance to review the "flip" PR and we merge it. (There are a few tweaks that I introduce wrt to datetime grids that I mostly want his approval on.)

**Comment by vincentarelbundock** [comment]
Converting to draft so I can move this closer to the ultimate goal before merging.

**Comment by vincentarelbundock** [comment]
This ended up being a massive PR, because I implemented custom types immediately. Sorry for the tough review!

(Note: this already includes the un-merged "flip" PR.)

Before looking at the code, I think we have to agree on the ergonomics for custom types. I added a vignette to the website and rendered it here:

https://arelbundock.com/tinyplot/vignettes/type.html

@grantmcdermott and @zeileis, if you could look at this vignette and weigh in on the ergonomics and concepts, that would be a great start.

The idea is:

* Every type is implemented by its own function: `"j"` is drawn by `type_jitter()`
* For each type, there are two operations:
    1. `data_jitter()` transforms `datapoints` and company.
    2. `draw_jitter()` calls graphics functions to draw the plot.

The benefits are:

1. Easy to define custom functions
2. Type-specific arguments do not "pollute" the main function. We can do:
    - `plt(y ~ x, type = type_jitter(amount = .1))`
3. Easier for users to find appropriate documentation:
    - `?type_jitter` instead of figuring out what gets pushed through `...`
4. Clear code organization in the repo

Let me know what you think.

**Comment by grantmcdermott** [comment]
Fantastic. Really excited by this functionality. I should have time to review this tomorrow, including design decisions, and hopefully Achim will have been able to give approval (or not!) of the flip PR, which will obviously simplify some things here.

**Comment by zeileis** [comment]
Wow, pretty impressive, Vincent @vincentarelbundock! I've just had a quick look at the basic design but I will try to implement a `type_spineplot()` 😇

**Comment by zeileis** [comment]
## Motivation

Vincent, thanks again! I was able to get started on `type_spineplot()` but I'm stuck on various aspects. Maybe there is already a good solution for this - if not, this might be helpful to think more about the design. There are two main sources of troubles: (1) Coercing factors to numeric an back. (2) Setting up the right kind of axes for the factor variables.

## Background

Spine plots are a type of stacked bar plots where the bar width is proportional to the relative proportion of each level of `x`. The bar heights are proportional to the conditional proportions of the levels of `y` within each `x` group. The overall height is always 1 = 100%. _(Hence the bars might resemble books on a shelf with the same height but different widths. Hence the name.)_

If `x` is numeric it is first grouped into bins (similar to a histogram) and then the same strategy as for categorical `x` is used. However, the axes annotation is then different. These plots are also called spinograms.

In summary, they are different, though, because both the x-axis and the y-axis have a cumulative probability scale between 0 and 1 (plus potential offsets). So it is not just 1, 2, ..., k where k is the number of categories.

## First spineplot attempt

In [type_spineplot.txt](https://github.com/user-attachments/files/17279072/type_spineplot.txt) there is a first implementation attempt of spineplots. The `data_spineplot()` is mostly copied from `data_boxplot()`. The `draw_spineplot()` copies some code from `graphics::spineplot.default()` but I tried to cut it down to a minimal version. I have excluded spinograms with numeric `x` for now and included a few further "TODO" things which I should eventually be able to sort out myself.

There are two "FIXME" comments in the code corresponding to my main issues above: handling of factors and axes.

## Example

I use the development version of `tinyplot`, the [type_spineplot.txt](https://github.com/user-attachments/files/17279072/type_spineplot.txt) code, and a slightly tweaked `airquality` data frame:

```
library("tinyplot")
source("type_spineplot.txt")
aq = transform(
  airquality,
  Month = factor(Month, labels = month.abb[unique(Month)]),
  Hot = Temp > median(Temp),
  Wind = cut(Wind, c(0, 5, 10, 15, 21))
)
```

Then first let's compare the default `tinyplot()` with the new `type` which does not get the `xlim`/`ylim` correct. Then we set these and it gets a lot better but the axis handling is still poor because we would need `yaxt` and `xaxt = "i"` and better labels. For comparison I include the standard `graphics::spineplot()` output:

```
tinyplot(Hot ~ Month, data = aq, type = type_spineplot())
tinyplot(Hot ~ Month, data = aq, type = type_spineplot(), xlim = c(0, 1.08), ylim = c(0, 1))
spineplot(factor(Hot) ~ Month, data = aq)
```

Spine plot of Hot vs. Month

Also note that the axis labeling will not necessarily be identical across facet groups because the conditional distribution of `x` may change as well:

```
tinyplot(Hot ~ Wind, facet = ~ Month, data = aq, type = type_spineplot(),
  xlim = c(0, 1.08), ylim = c(0, 1), facet.args = list(nrow = 1))
```

Spineplot of Hot vs. Wind given Month

## Summary

My impression is that we are already quite close to enabling spine plots and spinograms. The factor handling is probably already possible and I just didn't understand how to deal with the details. For the axis handling I'm not sure whether the current design is already general enough to enable everything that would be nice to have.

**Comment by vincentarelbundock** [comment]
This is super cool @zeileis 

I attach a simplified version of your code. Some stuff was removed for clarity and minimalism. Hopefully it shows you a path forward.

The two tricks I used:

* Instead of using `iy` and `ix`, I use the full `split_data` and the `i` index. This gives you access to the original factors, with levels, etc.
* `data_spineplot()` now returns a few more internal variables that will override whatever the user has supplied: `xlim`, `ylim`, `xaxt`, and `yaxt`.


[type_spineplot.txt](https://github.com/user-attachments/files/17282177/type_spineplot.txt)

Hopefully I understood what you needed and this helps.

**Comment by grantmcdermott** [comment]
Hey @vincentarelbundock I've merged the "flip" PR in main now. Sorry, but my changes have introduced a few conflicts that you'll need to merge in to this branch.

**Comment by vincentarelbundock** [comment]
Thanks for the ping @grantmcdermott . Conflicts should be fixed now.

**Comment by vincentarelbundock** [comment]
@zeileis 

I just pushed a change that should make these possible.

1. All the named arguments in the list that `data_spineplot()` returns override the variables in the main function scope. This means that you can already override `xaxs` and `yaxs` as needed by simply adding them to the output of `data_spineplot()`.
2. In the latest commit, I let `draw_spineplot()` access to a few more variables:
    - `data_by`: A list with all `datapoints` split along the `by` dimension (outter loop)
    - `data_facet`: A list with all `datapoints` split along the `facet` dimension (inner loop)
[type_spineplot.txt](https://github.com/user-attachments/files/17284511/type_spineplot.txt)

    - `iby`: Index the current `by` block that we are operating on.
    - `ifacet`: Index of the current `facet` block that we are operating on (within `by` block).

Attached is a working example with spineplots in multiple facets.

[type_spineplot.txt](https://github.com/user-attachments/files/17284522/type_spineplot.txt)

File: vignettes/type.qmd
Comment (grantmcdermott) [reply vignettes/type.qmd:1]: Oh, quasi-related: We do't have to address as part of this PR but once this is all merged then we probably want to revisit the prediction example(s) in the introductory vignette [here](https://github.com/grantmcdermott/tinyplot/blob/9da9e19f7977ef7eab6528dc2a72b21d5422f778/vignettes/intro_tutorial.qmd#L282-L313) with type = "lm".

And we should also link to this more detailed "types" vignette.

File: R/by_aesthetics.R
Comment (grantmcdermott) [inline R/by_aesthetics.R:1]: Sorry, I can see that a lot has been changed and moved around here, but can you give me the high-level summary of why these changes are necessary? I'm just trying to figure out how they relate to the elements refactor...

File: R/type_glm.R
Comment (vincentarelbundock) [reply R/type_glm.R:57]: This is so we can make GLM predictions on the link scale and transform them back to the response scale. Ensures confidence intervals are [0,1] in logit, for example. `backtransform` is just the argument name in the `ci()` function defined at the bottom of the `R/type_glm.R` file.

File: R/type_histogram.R
Comment (vincentarelbundock) [reply R/type_histogram.R:26]: No, that's different. In `tinyplot`, we filter `x`, `y`, `ymin`, etc. vectors to ensure that only non-empty vectors get bound in `datapoints`. In `type_histogram`, we filter out empty`by`/`facet` groups.

File: R/type_polypath.R
Comment (vincentarelbundock) [reply R/type_polypath.R]: 1e911d12dbca

File: R/type_glm.R
Comment (vincentarelbundock) [reply R/type_glm.R]: Need to use `stats::family` for the namespace benefit because there is also a `family` argument and we don't want collision.

File: vignettes/type.qmd
Comment (vincentarelbundock) [reply vignettes/type.qmd:1]: I kept the simple example but inserted a link to the Github repo where users can see all the `type_*()` functions we've implemented. Since there are a ton of complicated real-world examples there, I think that showing a truly minimalist case in the vignette serves a useful pedagogical purpose.

Have made the other suggested changes. a55af018d95f

File: R/type_boxplot.R
Comment (vincentarelbundock) [reply R/type_boxplot.R]: 374d5bc015c5

File: R/type_ribbon.R
Comment (vincentarelbundock) [reply R/type_ribbon.R:58]: 361aefae441d

File: R/draw_legend.R
Comment (vincentarelbundock) [reply R/draw_legend.R:169]: This is an example of what I got without this change: missing color boxes in the legend. I have not looked at the legend code enough to understand it, but this seemed to fix it. 

hist_grouped_fancy

File: R/sanitize.R
Comment (vincentarelbundock) [reply R/sanitize.R]: 39d29c2bc30a

File: R/tinyplot.R
Comment (vincentarelbundock) [reply R/tinyplot.R]: Turns out this one was useless. Rolled back: 10745ed1135a

File: R/tinyplot.R
Comment (vincentarelbundock) [reply R/tinyplot.R]: Re-instating. Ideally we would never change the value of `type` after running `sanitize_type()` at the very top. But fixing this bad practice would have ripple effects downstream and I don't have the energy to deal with it now. So reverting to the hack.

File: R/by_aesthetics.R
Comment (vincentarelbundock) [reply R/by_aesthetics.R:1]: There are basically no changes to that file. I appear to have run an autoformatter by mistake. Sorry. I reverted the indentation at least, so it's easier to check.

File: Makefile
Comment (vincentarelbundock) [reply Makefile:1]: FWIW, I do it the other way in `marginaleffects`: https://github.com/vincentarelbundock/marginaleffects/blob/main/Makefile

**Comment by vincentarelbundock** [comment]
@grantmcdermott issues all resolved (I think).

Happy to make more changes as you think is appropriate.

File: R/tinyplot.R
Comment (grantmcdermott) [inline R/tinyplot.R:1]: Just realized: We also need to document the new functional `<fun>_type()` options (and their character aliases) in the `type` argument.

File: vignettes/intro_tutorial.qmd
Comment (grantmcdermott) [inline vignettes/intro_tutorial.qmd:332]: Could you perhaps add a sentence or two along the lines of:

"We'll see other examples of additional plot types below, including a few more of the special model prediction types.  But please also take a look at the dedicated [Plot types](https://grantmcdermott.com/tinyplot/vignettes/types.html) vignette for explicit details on all of the different plot types that **tinyplot** supports, as well as how to roll your own custom types."

**Review by grantmcdermott — CHANGES_REQUESTED** [review]
Excellent, thanks @vincentarelbundock. I think we're close now. I left a few comments that I missed first time around and also have two meta requests:

1. Can you please add some snapshot tests for the new model  prediction types (lm/glm/loess)?
2. Can you please add a NEWS item. Don't forget to credit yourself!

**Comment by vincentarelbundock** [comment]
Snapshots: 7277ceee8bcd

Type documentation: d9be3388bd59

Vignette improvements: 2d9df69b8fd2

News: b328e950b3c7

File: vignettes/intro_tutorial.qmd
Comment (grantmcdermott) [inline vignettes/intro_tutorial.qmd:355]: With the advent of `type = type_lm()` /  `type = "lm"` here, we should probably convert these examples to the formula syntax for ease of readability and consistency with the preceding examples. I'm happy to do that separately (i.e., after merging); just making a mental note for myself.

File: vignettes/intro_tutorial.qmd
Comment (grantmcdermott) [inline vignettes/intro_tutorial.qmd:360]: Q: Do you perfer to use `type = type_lm()` rather than `type = "lm"`. I don't have strong preferences, but lean slightly toward the former since I suspect users will find it cool that they can use the character shortcut. (Your call; I just wanted to flag.)

**Review by grantmcdermott — CHANGES_REQUESTED** [review]
_Nearly_ there. Only a few more comments. Thanks for bearing with me @vincentarelbundock.

**Comment by vincentarelbundock** [comment]
All good suggestions. Resolved all, but the formula conversion. Agree but have other things to attend now. Might circle back eventually if I think about it.

**Comment by grantmcdermott** [comment]
Brilliant. I'm happy to merge now, but will change the PR title first if you don't mind (to something more explicit given the expanded scope of the PR).

**Review by grantmcdermott — APPROVED** [review]
Outstanding @vincentarelbundock! Thanks so much for putting in the effort, as well as the careful design consideration.

Really, really stoked with this functionality.

**Pull request merged**
